Problem: tests fail on Windows

yrashk opened the pull request:
Also, if a non-exe file maching binary is present in the
build directory, it'll get picked up but there'll be an error
saying that this binary is not a valid Win32 binary.

Solution: adjust tests to accommodate for Windows' ecosystem
and make the binary name automatically assume .exe extension
when compiled on Windows.

emk commented on cli_test_dir/src/lib.rs:
```diff
@@ -42,17 +42,29 @@
+//! #[cfg(unix)]
 //! let testdir = TestDir::new("true", "true_succeeds");
-//! testdir.cmd().expect_success();
+//! #[cfg(windows)]
```
Thank you for trying to fix the Windows tests!

However, since this is documentation, I would strongly prefer to avoid filling the API docs with `#[cfg(...)]` statements. For the public-facing documentation, we might hide the `#[cfg(...)]` machinery as follows:

```rust
//! ```
//! # use cli_test_dir::*;
//! # #[cfg(unix)]
//! let testdir = TestDir::new("true", "true_succeeds");
//! # #[cfg(windows)]
//! # let testdir = TestDir::new("cmd", "true_succeeds"); 
//! testdir.cmd().expect_success();
//! ```
```

...but I'd accept a more elegant solution if you can think of one. For example, there might be shell commands which are identical on both platforms.

emk commented on cli_test_dir/src/lib.rs:
```diff
@@ -205,7 +252,10 @@ impl TestDir {
+        #[cfg(unix)]
         let mut bin = bin_dir.join(&bin_name);
+        #[cfg(windows)]
+        let mut bin = bin_dir.join(format!("{}.exe", bin_name));
```
Yes, we should definitely fix this!

However, I would personally prefer to apply `#[cfg(...)]` to entire functions instead of individual statements. (I don't know why I prefer to do things this way.) Could be we perhaps add a private `exe_name(name: &str) -> Cow<str>` function which implements this?

```rust
use std::borrow::Cow;

#[cfg(unix)]
fn exe_name(name: &str) -> Cow<str> {
    Cow::Borrowed(name)
}

#[cfg(windows)]
fn exe_name(name: &str) -> Cow<str> {
    // Maybe something like...
    if name.ends_with(".exe") {
        Cow::Borrowed(name)
    } else {
        Cow::Owned(format!("{}.exe", name))
    }
}
```

emk reviewed (changes requested):
Thank you for submitting a PR to fix this issue! I'd love for this library to work well on Windows.

I have a couple of minor suggestions, though. :-)

yrashk committed:
Problem: tests fail on Windows

Also, if a non-exe file maching binary is present in the
build directory, it'll get picked up but there'll be an error
saying that this binary is not a valid Win32 binary.

Solution: adjust tests to accommodate for Windows' ecosystem
and make the binary name automatically assume .exe extension
when compiled on Windows.

yrashk commented:
@emk I've implemented changes you requested

emk commented:
Thank you for this PR and for making the changes! Merging.

yrashk commented:
Thank you! Will you be able to publish a new create as well?

On Fri, Apr 20, 2018, 7:30 AM Eric Kidd <notifications@github.com> wrote:

> Merged #43 <https://github.com/emk/subtitles-rs/pull/43>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/emk/subtitles-rs/pull/43#event-1586301636>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAABxCMWJTpZk3YzGUQluWfxQ3Q3e6V_ks5tqfD5gaJpZM4TcUM->
> .
>

emk commented:
If I don't publish a new crate this weekend, ping me please.

yrashk commented:
Ok, I will. In the meantime I will publish an interim fork crate, I suppose.

emk commented:
I'm not sure why failing unit tests on Windows for an extra ~24 hours is worth publishing a fork crate? Why not just use `{ git = "...", branch = "..." }` for an extra day?

`cli_test_dir` is actually very actively used (it's part of a lot of internal stuff at work), but it's hard for me to turn around PRs in under 48 hours when I'm traveling.

yrashk commented:
Eric, no worries at all. Just looking at maintaining a piece of software
that doesn't depend on git versions as these are hard to track and there
are some environments that don't like git sources (such as NixOS' nixpkgs).

I'll either just postpone my own update of the package or do a fork. The
former is more likely anyway given the weekend :)

On Fri, Apr 20, 2018, 4:45 PM Eric Kidd <notifications@github.com> wrote:

> I'm not sure why failing unit tests on Windows for an extra ~24 hours is
> worth publishing a fork crate? Why not just use { git = "...", branch =
> "..." } for an extra day?
>
> cli_test_dir is actually very actively used (it's part of a lot of
> internal stuff at work), but it's hard for me to turn around PRs in under
> 48 hours when I'm traveling.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/emk/subtitles-rs/pull/43#issuecomment-383249028>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAABxFZc_OXVEe3b0yJCCi_s3GxUEbC_ks5tqnMTgaJpZM4TcUM->
> .
>

emk commented:
I've released 0.1.5! (It took slightly longer than planned because of some home maintenance projects this weekend.)

Thank you very much for submitting this PR and for your patience!
